File: Python/M96_UniqueBinarySearchTrees.py
from functools import reduce #python3 
class Solution:
    def fact(self, n):
        return reduce(lambda x,y: x*y, [1] + list(range(1,n+1)))
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        #calculate catalan number
        fcn = self.fact(n)
        return self.fact(2*n) // (fcn * fcn * (n+1))
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        r = [1,1]
        for i in range(2,n+1):
            ix = i - 1
            t = ((4*ix+2) *r[ix])//(ix+2)
            r.append(t)
        return r[n]
        
        
       
def main():
    n = 4
    s = Solution().numTrees(n)
    print(s)

if __name__=='__main__':
    main()

File: LeetCode/Cpp/M96_UniqueBinarySearchTrees.cpp
class Solution {
public:
    vector<long long> v;
    double fact(int n){
        double d=1;
        for(int i=1;i<=n;i++){
            d *= i;
            
        }
        return d;
    }
    int numTrees(int n) {
        return fact(2*n)/(fact(n+1)*fact(n))+0.5;
    }
};

int main(){
    int n = 3;
    s = Solution();
    int ret = s.numTrees(n);
    printf("%d",ret);
    return 0;
}